airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java

```diff
@@ -0,0 +1,218 @@
+          min.map(Version::serialize).orElse(""), max.map(Version::serialize).orElse(""));
+    }
+
+    return new AirbyteProtocolVersionRange(min.orElse(AirbyteProtocolVersion.DEFAULT_AIRBYTE_PROTOCOL_VERSION),
+        max.orElse(AirbyteProtocolVersion.DEFAULT_AIRBYTE_PROTOCOL_VERSION));
+  }
+
+  protected AirbyteProtocolVersionRange getTargetProtocolVersionRange() {
+    return new AirbyteProtocolVersionRange(configs.getAirbyteProtocolVersionMin(), configs.getAirbyteProtocolVersionMax());
+  }
+
+  protected Map<ActorType, Set<UUID>> getConflictingActorDefinitions(final AirbyteProtocolVersionRange targetRange) throws IOException {
+    final Map<UUID, Map.Entry<ActorType, Version>> actorDefIdToProtocolVersion = configRepository.getActorDefinitionToProtocolVersionMap();
+    final Map<ActorType, Set<UUID>> conflicts =
+        actorDefIdToProtocolVersion.entrySet().stream()
+            // Keeping only ActorDefinitionIds that have an unsupported protocol version
+            .filter(e -> !targetRange.isSupported(e.getValue().getValue()))
+            // Build the ActorType -> List[ActorDefIds] map
+            .map(e -> Map.entry(e.getValue().getKey(), e.getKey()))
+            // Group by ActorType and transform the List<Entry<ActorType, UUID>> into a Set<UUID>
+            .collect(Collectors.groupingBy(Entry::getKey,
+                Collectors.collectingAndThen(Collectors.toList(), list -> list.stream().map(Entry::getValue).collect(Collectors.toSet()))));
+    return conflicts;
+  }
+
+  protected Set<UUID> projectRemainingConflictsAfterConnectorUpgrades(final AirbyteProtocolVersionRange targetRange,
+                                                                      final Set<UUID> initialConflicts,
+                                                                      final ActorType actorType) {
+    if (initialConflicts.isEmpty()) {
+      return Set.of();
+    }
+
+    final Set<UUID> upgradedSourceDefs = getProtocolVersionsForActorDefinitions(actorType)
+        // Keep definition ids if the protocol version will fall into the new supported range
+        .filter(e -> initialConflicts.contains(e.getKey()) && targetRange.isSupported(e.getValue()))
+        .map(Entry::getKey)
+        .collect(Collectors.toSet());
+
+    // Get the set of source definitions that will still have conflict after the connector upgrades
+    final Set<UUID> remainingConflicts = new HashSet<>(initialConflicts);
+    remainingConflicts.removeAll(upgradedSourceDefs);
+    return remainingConflicts;
+  }
+
+  protected Stream<Entry<UUID, Version>> getProtocolVersionsForActorDefinitions(final ActorType actorType) {
+    Stream<Entry<UUID, Version>> stream;
+    if (actorType == ActorType.SOURCE) {
+      stream = definitionsProvider.getSourceDefinitions()
+          .stream()
+          .map(def -> Map.entry(def.getSourceDefinitionId(), AirbyteProtocolVersion.getWithDefault(def.getSpec().getProtocolVersion())));
```

Comment:
Why not doing collect(Collectors.toMap) here? or just map the to get the keys. It looks like this is the only thing that is neeed here.